controllers/rollout/batchrelease/inner_controller.go

```diff
@@ -0,0 +1,263 @@
+		Paused:               batch.Spec.ReleasePlan.Paused,
+	}
+	if batch.Status.CanaryStatus.CurrentBatchState == appsv1alpha1.ReadyBatchState {
+		state.State = BatchReadyState
+	} else {
+		state.State = BatchInRollingState
+	}
+	return state, nil
+}
+
+func (r *innerBatchController) PromoteBatch(index int32) error {
+	index = index - 1
+	batch := &appsv1alpha1.BatchRelease{}
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := r.Get(context.TODO(), client.ObjectKey{Namespace: r.rollout.Namespace, Name: r.batchName}, batch); err != nil {
+			klog.Errorf("error getting updated rollout(%s/%s) from client", batch.Namespace, batch.Name)
+			return err
+		}
+		if !batch.Spec.ReleasePlan.Paused && *batch.Spec.ReleasePlan.BatchPartition == index {
+			return nil
+		}
+		batch.Spec.ReleasePlan.BatchPartition = utilpointer.Int32Ptr(index)
+		batch.Spec.ReleasePlan.Paused = false
+		if err := r.Client.Update(context.TODO(), batch); err != nil {
+			return err
+		}
+		klog.Infof("rollout(%s/%s) promote batchRelease(%s) BatchPartition(%d) success", r.rollout.Namespace, r.rollout.Name, r.batchName, index)
+		return nil
+	}); err != nil {
+		klog.Errorf("rollout(%s/%s) promote batchRelease(%s) BatchPartition(%d) failed: %s", r.rollout.Namespace, r.rollout.Name, r.batchName, index, err.Error())
+		return err
+	}
+	return nil
+}
+
+func (r *innerBatchController) ResumeStableWorkload(checkReady bool) (bool, error) {
+	// todo cloneSet
+	if r.rollout.Spec.ObjectRef.WorkloadRef.Kind == util.ControllerKruiseKindCS.Kind {
+		return true, nil
+	}
+
+	// deployment
+	dName := r.rollout.Spec.ObjectRef.WorkloadRef.Name
+	obj := &apps.Deployment{}
+	err := r.Get(context.TODO(), types.NamespacedName{Namespace: r.rollout.Namespace, Name: dName}, obj)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Warningf("rollout(%s/%s) stable deployment(%s) not found, and return true", r.rollout.Namespace, r.rollout.Name, dName)
+			return true, nil
+		}
+		return false, err
+	}
+	// set deployment paused=false
+	if obj.Spec.Paused == true {
+		err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			if err = r.Get(context.TODO(), types.NamespacedName{Namespace: r.rollout.Namespace, Name: dName}, obj); err != nil {
+				return err
+			}
+			obj.Spec.Paused = false
+			return r.Update(context.TODO(), obj)
+		})
+		if err != nil {
+			klog.Errorf("update rollout(%s/%s) stable deployment failed: %s", r.rollout.Namespace, r.rollout.Name, err.Error())
+			return false, err
+		}
+		klog.Infof("resume rollout(%s/%s) stable deployment(paused=false) success", r.rollout.Namespace, r.rollout.Name)
+		return false, nil
+	}
+
+	// Whether to wait for pods are ready
+	if !checkReady {
+		return true, nil
+	}
+	by, _ := json.Marshal(obj.Status)
+	// wait for all pods are ready
+	maxUnavailable, _ := intstr.GetValueFromIntOrPercent(obj.Spec.Strategy.RollingUpdate.MaxUnavailable, int(*obj.Spec.Replicas), true)
```

Comment:
GetValueFromIntOrPercent was deprecated in favor of GetScaledValueFromIntOrPercent